modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java

```diff
@@ -603,62 +608,60 @@ void run(BytesReference expected, BytesReference updated, ActionListener<Optiona
-                            )
-                        ),
-                        doCleanup
-                    )
-                )
-            ) {
-                if (currentUploads.size() > 1) {
+                // Step 3: Cancel all the other uploads against which we are racing, leaving some time if we are not first in the queue to
+                // give them some time to complete.
+
+                .<Void>newForked(otherUploadsCancelledListener -> {
                     // This is a small optimization to improve the liveness properties of this algorithm.
                     //
                     // When there are multiple competing updates, we order them by upload id and the first one tries to cancel the competing
                     // updates in order to make progress. To avoid liveness issues when the winner fails, the rest wait based on their
                     // upload_id-based position and try to make progress.
 
-                    var delayListener = listeners.acquire();
-                    final Runnable cancelConcurrentUpdates = () -> {
-                        try {
-                            cancelOtherUploads(uploadId, currentUploads, listeners);
-                        } finally {
-                            delayListener.onResponse(null);
-                        }
-                    };
-
                     if (uploadIndex > 0) {
                         threadPool.scheduleUnlessShuttingDown(
                             TimeValue.timeValueMillis(
                                 uploadIndex * blobStore.getCompareAndExchangeAntiContentionDelay().millis() + Randomness.get().nextInt(50)
                             ),
                             blobStore.getSnapshotExecutor(),
-                            cancelConcurrentUpdates
+                            ActionRunnable.wrap(otherUploadsCancelledListener, l -> cancelOtherUploads(uploadId, currentUploads, l))
                         );
                     } else {
-                        cancelConcurrentUpdates.run();
+                        cancelOtherUploads(uploadId, currentUploads, otherUploadsCancelledListener);
                     }
-                }
-            }
+                })
+
+                // Step 4: Read the current register value.
+
+                .<OptionalBytesReference>andThen((l, ignored) -> getRegister(purpose, rawKey, l))
+
+                // Step 5: Perform the compare-and-swap by completing our upload iff the witnessed value matches the expected value.
+
+                .<OptionalBytesReference>andThen((l, currentValue) -> ActionListener.completeWith(l, () -> {
+                    if (currentValue.isPresent() && currentValue.bytesReference().equals(expected)) {
+                        completeMultipartUpload(uploadId, partETag);
+                    } else {
+                        // Best-effort attempt to clean up after ourselves.
+                        safeAbortMultipartUpload(uploadId);
```

Comment:
This now runs in the thread before we return `currentValue`. Current code has it wrapped inside a `runAfter`. Do we want to retain this behaviour?

Reply:
I think it's equivalent, similarly to how

```java
    try {
        return x;
    } finally {
        foo();
    }
```

is semantically identical to

```java
    foo();
    return x;
```

since `return x` cannot throw.

Reply:
OK you are right. I forgot `runAfter` is done with try-finally. I thought it was:
```
listener.onResponse(x)
cleanup();
```
Apparently it does _not_ make sense as a generic run-after implementation since `onReponse` may throw.

Reply:
Ah sorry you're quite right, before this change we would complete the listener _and then_ do the cleanup action. That seems undesirable to me, ideally we should finish doing all nontrivial work before responding to the caller. The code as written is what I wanted in the first place.

Reply:
Ah right, `listener.onResponse` is not a `return` so it does execute first. If the change here is intentional, that's all good. Thanks!